### ISSUE TITLE
How can I run some part of code only once before all spec files?

### ISSUE THREAD

**Issue by sanyco92** [webdriverio/webdriverio#3747]
Hello guys. Need help. I have 2 spec files (2 test suites, but both files actually gathered into 1 test suite). I need to execute some actions only once before 2 specs, (e.g. run the mobile application and download all available files in the application), and then continue run test specs one by one. I tried to use 'before' hook, but unfortunately, it's being executed before each spec file. Is there any way to run a part of code only once before all spec files?

The suite is described in the config file like this:
```
suites: {
        smoke: [
            './specs/smoke/dcm.spec.js',
            './specs/smoke/forms.spec.js'
        ],
    },
```

I'm running it like this:
`wdio ./config/platforms/ios/wdio.conf.ios.ipad.air.js --suite smoke`

**Environment (please complete the following information):**
 - **WebdriverIO version:** [4.14.1]
 - **Mode:** [WDIO Testrunner]
 - **If WDIO Testrunner, running sync/async:** [sync]

**Comment by christian-bromann** [comment]
Set `maxInstances: 1`

in future ... join our  [![Gitter](https://badges.gitter.im/webdriverio/webdriverio.svg)](https://gitter.im/webdriverio/webdriverio?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge) for these kind of questions. This is not necessary an issue.

**Comment by sanyco92** [comment]
I have this option in my config file, but it doesn't help. 
I asked for help in gitter, but no one answered. 
And I didn't open the issue, I selected option: "ask question" and it still not resolved. Could you open it please?

**Comment by christian-bromann** [comment]
please use the onPrepare hook if you only want to do things once, see also docs (https://webdriver.io/docs/configurationfile.html):

```js
    /**
     * Gets executed once before all workers get launched.
     * @param {Object} config wdio configuration object
     * @param {Array.<Object>} capabilities list of capabilities details
     */
    onPrepare: function (config, capabilities) {
    },
```

**Comment by sanyco92** [comment]
Thanks for your answer, but I think it also will not help. I need to do actions in mobile application using browser variable, but in this hook I don't have access to it.

**Comment by mgrybyk** [comment]
@sanyco92 if you add your code to before block and only trigger it based on some condition, ex: some global.isSomethingDone?

**Comment by mgrybyk** [comment]
```if !global.issometing
  do something
  global.issomething = true
```

**Comment by christian-bromann** [comment]
@mgrybyk workers run in different processes so no chance to access a global variable between tests.

@sanyco92 please understand that it is impossible to answer your question given the details you have given. Usually all tests have to run independently from each other so having to do one thing that applies for all tests but only when all sessions have been established sounds like a bad way of testing things. Again I am not complete aware of the situation you are facing. Therefor I recommend to reach out to the support channel with specific questions. Or make specific feature requests.

**Comment by sanyco92** [comment]
@christian-bromann ok, I'll try to give more detailed info. I have a mobile application, which consists of ~6-7 different modules with some functionality. For each module, I'm developing some tests (it will be around 100 tests for each module) and these tests I'll store in spec file for each module. So, in the future, there will be ~7 spec files with ~100 tests in each. But, there's one more thing. Each module has some environment (pdf files, HTML files, audio/video files, different documents, tables, forms), and all these files are syncing with the user in the app and downloading when the application is launched for the first time and user logged in. This downloading process takes around 1 hour, and right now, when I have only 2 spec files, I'm spending 2 hours for downloading. All I want is to download all these files only one time before all spec files are launched. So I need just to install the application, sign in, make sync and download all required files for tests (only once before all tests and before all spec files), and then launch spec files one by one (without reinstalling the application and without clearing the application data).

I have 2 config files, that are merged into one config file file.

master.conf.js:
```
exports.config = {

    specs: [
        './specs/*/*.spec.js'
    ],

    exclude: [
        //
    ],

    suites: {
        nightly: [
            './specs/nightly/1.spec.js',
            './specs/nightly/2.spec.js',
            './specs/nightly/3.spec.js',
            './specs/nightly/4.spec.js',
            './specs/nightly/5.spec.js',
            './specs/nightly/6.spec.js',
            './specs/nightly/7.spec.js',
            './specs/nightly/8.spec.js',
            './specs/nightly/9.spec.js',
        ],

        smoke: [
            './specs/smoke/module1.spec.js',
            './specs/smoke/module2.spec.js'
        ],
    },

    sync: true,
    logLevel: 'verbose',
    coloredLogs: true,
    deprecationWarnings: false,
    bail: 0,
    baseUrl: 'http://localhost',
    waitforTimeout: 90000,
    connectionRetryTimeout: 240000,
    connectionRetryCount: 3,
    framework: 'mocha',
    reporters: ['spec', 'allure'],

    mochaOpts: {
        ui: 'bdd',
        timeout: 3600000,
    },

    before: function (capabilities, specs) {
        let chai = require("chai");
        chai.config.includeStack = true;
        chai.config.showDiff = true;
        chai.config.truncateThreshold = 0;
        global.expect = chai.expect;
    },
};
```

wdio.conf.js:
```
let devConf = {

    host: 'localhost',

    port: 4444,

    capabilities: [{
        maxInstances: 1,
        platformName: 'iOS',
        platformVersion: '12.1',
        deviceName: 'iPhone 8',
        automationName: 'XCUITest',
        app: path.resolve("./*******/***.app"),
        showXcodeLog: false,
        launchTimeout: 600000,
        clearSystemFiles: true,
        wdaStartupRetries: 3000,
        wdaStartupRetryInterval: 1000,
        useNewWDA: true,
        shouldWaitForQuiescence: false,
        shouldUseSingletonTestManager: false,
        wdaLaunchTimeout: 2400000,
        wdaConnectionTimeout: 2400000,
        newCommandTimeout: 600000,
        noReset: false,
        maxTypingFrequency: 40,
        wdaLocalPort: 8100,
    }],

    services: ['appium'],

    appium: {
        // waitStartTime: 120000,
        logFileName: 'appium.log',
        args: {
            address: 'localhost',
            port: 4444,
            commandTimeout: '7200',
            sessionOverride: true,
            debugLogSpacing: true
        }
    },

    beforeSuite: function (suite) {
        let username = credentials.username,
            password = credentials.password,

        loginPage.acceptAlert();
        loginPage.loginWithCredentials(username, password);
        homePage.waitForDownloadsComplete(); // waiting for all files being downloaded 
        helpers.restartApp(); // here I'm closing and opening the application (without reinstall)
    },

    beforeTest: function (test) {
        switch (test.parent) {
            case 'Module1':
                homePage.openModule(homePage.moduleOne);
                break;
            case 'Module2':
                homePage.openModule(homePage.moduleTwo);
                break;
        }
    },

    afterTest: function (test) {
        if (!test.passed) {
            helpers.takeScreenshot(test);
        }
    },
};

exports.config = merge.all([wdioConf.config, devConf]);
```

I'm running tests, as described above, like this:
wdio ./config/platforms/ios/wdio.conf.ios.ipad.air.js --suite smoke

It launches 2 spec files from 'smoke' suite and 'before' hook is being executed for each spec file, the downloading process is started before each spec file.  I want to reduce tests run time by executing the download process only once, I don't need to reinstall the application and download files for each spec file. Thank you in advance and hope this info will help for a better understanding of my problem.

**Comment by CrispusDH** [comment]
@sanyco92 you can write script that will be do whatever you need. For example `before-tests-hook.js`. And add it to your `package.json` scripts like:
```js
    "hook:before": "node dist/scripts/before-tests-hook.js",
    "pretest": "npm run build && npm run hook:before",
    "test": "npm run .....",
```
I think you got idea:
- create script
- run it when you want

**Comment by sanyco92** [comment]
> @sanyco92 you can write script that will be do whatever you need. For example `before-tests-hook.js`. And add it to your `package.json` scripts like:
> 
> ```js
>     "hook:before": "node dist/scripts/before-tests-hook.js",
>     "pretest": "npm run build && npm run hook:before",
>     "test": "npm run .....",
> ```
> 
> I think you got idea:
> 
> * create script
> * run it when you want

Think it can be the solution. Thanks!

**Comment by chanduguthikonda** [comment]
sanyco92 I know I'm starting conversation on very old thread. I'm actually new to wdio and java script. I have a similar requirement to execute piece of code before executing each spec. So I was trying to call and execute login spec from beforeSuite: function in the below manner.

login.test.js:

`import { expect } from 'chai';  // Using Expect style.
import LoginPage from '../../pageobjects/login.page';
import WatchNowPage from '../../pageobjects/watch-now.page';
describe('Login Page Tests', function () {
  context('when on Login Page', () => {
    it('displays logo', function () {
      browser.waitUntil(
        () => LoginPage.loginLogo.getAttribute("isdisplayed") === true,
        10000, 'expected text to be different after 10s'
      );
    expect(LoginPage.loginLogo.getAttribute('isdisplayed')).to.be.true;
    });
  }
}`

wdio.conf.js: I need to execute the above spec from beforeSuite:

`beforeSuite: function (suite) {
    const testlogin = require('./test/specs/login/login.test.js');
    testlogin.run; //NOT sure if this works
  },`

From the code you mentioned above, I see you are calling functions on login page in beforeSuite function. Thats my actual requirement.
Any help would be highly appreciated
